torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
   }
 }
 
+bool shouldConsiderForMerge(detail::Vertex<Node*>* v) {
+  if (v->rdata.size() >= 2) {
+    return true;
+  }
+  JIT_ASSERT(v->rdata.size() == 1);
+  auto * node = *v->rdata.begin();
+  if (node->kind() == prim::Constant) {
+    return false;
+  }
+  return isDifferentiable(node);
+}
+
+static detail::DynamicDAG<Node*> make_dependency_graph(Block * block) {
+  detail::DynamicDAG<Node*> dag;
+  std::unordered_map<Node*,detail::Vertex<Node*>*> node_to_vertex;
+  // NB: the block's param and return nodes are not in the dependency graph.
+  for (Node * node : block->nodes()) {
+    node_to_vertex[node] = dag.newVertex(node);
+  }
+  for (auto * node : block->nodes()) {
+    for (auto * v : node->outputs()) {
+      for (auto & use : v->uses()) {
+        // [Determine data dependencies]
+        // Consider the following code:
+        //     y = f(x)
+        //     if k:
+        //        w += y
+        //     z = g(y)
+        // This produces a dependency graph with 3 vertices:
+        // (0: f)   (1: if k ...)   (2: g)
+        // We need to peek into the if Node* to determine its data dependencies
+        // (the body depends on the output of f, so Vertex 1 depends on Vertex 0).
+        // For each Use of y, we find an owning node of y that is a part of the
+        // dependency graph (in this case, the Vertex containing the if Node*)
+        // and then record the dependency.
+        auto * owning_node = use.user;
+        while (owning_node != nullptr) {
+          auto search = node_to_vertex.find(owning_node);
+          if (search == node_to_vertex.end()) {
+            owning_node = owning_node->owningBlock()->owningNode();
+            continue;
+          }
+          dag.addEdge(node_to_vertex[node], search->second);
+          break;
+        }
+      }
+    }
+  }
+  return dag;
+}
+
+static void find_differentiable_groups(
+    detail::DynamicDAG<Node*>& dep_graph,
+    size_t distance_threshold=64,
+    size_t producer_outedge_threshold=16) {
+  // A Vertex contains a Node* or a differential group of Node*.
+  // Perform graph contraction on dep_graph: contract two vertices(x, y) if
+  // the following conditions hold:
+  // - x, y can be merged to form a differential group
+  // - the contraction would not invalidate the dag (it creates no cycles).
+
+  // Iterate in reverse topological order
+  int64_t ord = dep_graph.maybe_vertices().size() - 1;
+  while (ord >= 0) {
+    if (!dep_graph.maybe_vertices().at(ord)) {
+      --ord;
```

Comment:
We decrement this on 3 paths, and want to keep its value on a single path. Maybe let's write this thing as a `for` loop which decrements, and simply do `++ord` in that single place that actually wants to stay at the same index (with a comment)?